File: tonghua-project/backend/app/schemas/supply_chain.py
from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class SupplyChainRecordCreate(BaseModel):
    product_id: int = Field(..., description="Associated product ID")
    stage: str = Field(
        ...,
        pattern="^(material_sourcing|processing|manufacturing|quality_check|shipping)$",
        description="Supply chain stage: material_sourcing, processing, manufacturing, quality_check, shipping",
    )
    description: Optional[str] = Field(None, description="Stage description and details")
    location: Optional[str] = Field(None, max_length=300, description="Geographic location of this stage")
    certified: bool = Field(False, description="Whether this stage has certification")
    cert_image_url: Optional[str] = Field(None, max_length=500, description="Certification document image URL")
    timestamp: Optional[datetime] = Field(None, description="Actual date/time of this stage")


class SupplyChainRecordUpdate(BaseModel):
    description: Optional[str] = None
    location: Optional[str] = Field(None, max_length=300)
    certified: Optional[bool] = None
    cert_image_url: Optional[str] = Field(None, max_length=500)
    timestamp: Optional[datetime] = None


class SupplyChainRecordOut(BaseModel):
    id: int
    product_id: int
    stage: str
    description: Optional[str] = None
    location: Optional[str] = None
    certified: bool
    cert_image_url: Optional[str] = None
    timestamp: Optional[datetime] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class SupplyChainTrace(BaseModel):
    """Full supply chain trace for a product."""
    product_id: int
    product_name: str
    records: List[SupplyChainRecordOut] = []

File: tonghua-project/frontend/web-react/src/services/supply-chain.ts
import api from './api';

export interface SupplyChainRecord {
  id: number;
  product_id: number;
  stage: string;
  description: string;
  location: string;
  certified: boolean;
  cert_image_url: string | null;
  timestamp: string;
  created_at: string;
}

export interface SupplyChainTrace {
  product_id: number;
  product_name: string;
  records: SupplyChainRecord[];
}

export interface SupplyChainStage {
  key: string;
  label: string;
  order: number;
}

export const supplyChainApi = {
  getRecords: async (params?: {
    product_id?: number;
    stage?: string;
    page?: number;
    page_size?: number;
  }): Promise<{ items: SupplyChainRecord[]; total: number; page: number; page_size: number }> => {
    const response = await api.get('/supply-chain/records', { params });
    const paginated = response.data;
    return {
      items: paginated.data ?? [],
      total: paginated.total ?? 0,
      page: paginated.page ?? 1,
      page_size: paginated.page_size ?? 20,
    };
  },

  trace: async (productId: string | number): Promise<SupplyChainTrace> => {
    const response = await api.get(`/supply-chain/trace/${productId}`);
    return response.data.data;
  },

  getStages: async (): Promise<SupplyChainStage[]> => {
    const response = await api.get('/supply-chain/stages');
    return response.data.data ?? [];
  },
};
